upload-api/billing.js

```diff
@@ -0,0 +1,89 @@
+import { Failure } from '@ucanto/core'
+import { toEmail } from '@web3-storage/did-mailto'
+
+export class InvalidSubscriptionState extends Failure {
+  /**
+   * @param {string} [message] Context for the message.
+   * @param {ErrorOptions} [options]
+   */
+  constructor(message, options) {
+    super(undefined, options)
+    this.name = /** @type {const} */ ('InvalidSubscriptionState')
+    this.detail = message
+  }
+
+  describe() {
+    return `subscription cannot be updated because it is not in a valid state: ${this.detail}`
+  }
+}
+
+export class BillingProviderUpdateError extends Failure {
+  /**
+   * @param {string} [message] Context for the message.
+   * @param {ErrorOptions} [options]
+   */
+  constructor(message, options) {
+    super(undefined, options)
+    this.name = /** @type {const} */ ('BillingProviderUpdateError')
+    this.detail = message
+  }
+
+  describe() {
+    return `encountered an error updating subscription: ${this.detail}`
+  }
+}
+
+/**
+ * 
+ * @param {import('stripe').Stripe} stripe 
+ * @returns {import("./types").BillingProvider}
+ */
+export function createStripeBillingProvider(stripe) {
+  return {
+    async hasCustomer(customer) {
+      const customersResponse = await stripe.customers.list({ email: toEmail(/** @type {import('@web3-storage/did-mailto').DidMailto} */(customer)) })
```

Comment:
this layer should probably *normalize* these email addresses before reading/writing.

e.g. lowercase the domain part of the email address before passing to stripe.
https://stackoverflow.com/questions/9807909/are-email-addresses-case-sensitive

(ideally stripe would do this for you but I wouldn't count on it)

Reply:
I wonder if we should also normalize the part before the `@` for email providers (like gmail, outlook and yahoo) that treat it as case-insensitive: https://www.linkedin.com/pulse/unraveling-mystery-email-addresses-case-sensitive-warmy#:~:text=Gmail's%20Approach%20to%20Case%20Sensitivity.&text=This%20means%20that%20regardless%20of,email%20address%20in%20Gmail's%20system.

for gmail I think this would also mean stripping out `.` from that part of the email address: https://support.google.com/mail/answer/7436150?hl=en